File: 1/2020.11.01/1122.Relative Sort Array/python/solution.py
class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        # bucket sort
        bucket = [0 for _ in range(1001)]
        # put arr1's elements in bucket
        for num in arr1:
            bucket[num] += 1
        res = []
        for num in arr2:
            for _ in range(bucket[num]):
                res.append(num)
                bucket[num] -= 1
        for i in range(len(bucket)):
            for _ in range(bucket[i]):
                res.append(i)
        return res

File: 1/2020.11.01/1122.Relative Sort Array/java/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        // bucket sort
        int[] bucket = new int[1001];
        for (int num : arr1) {
            bucket[num]++;
        }
        // use arr1 itself
        int index = 0;
        for (int num : arr2) {
            while (bucket[num]-- > 0) {
                arr1[index++] = num;
            }
        }
        for (int i = 0; i < bucket.length; i++) {
            while (bucket[i]-- > 0) {
                arr1[index++] = i;
            }
        }
        return arr1;
    }
}
